pkg/inject/inject.go

```diff
@@ -0,0 +1,585 @@
+package inject
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/linkerd/linkerd2/controller/gen/config"
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	log "github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	k8sMeta "k8s.io/apimachinery/pkg/api/meta"
+	k8sResource "k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	// LocalhostDNSNameOverride allows override of the controlPlaneDNS. This
+	// must be in absolute form for the proxy to special-case it.
+	LocalhostDNSNameOverride = "localhost."
+	// ControlPlanePodName default control plane pod name.
+	ControlPlanePodName = "linkerd-controller"
+	// PodNamespaceEnvVarName is the name of the variable used to pass the pod's namespace.
+	PodNamespaceEnvVarName = "LINKERD2_PROXY_POD_NAMESPACE"
+	// PrometheusImage is the docker image and tag for the Prometheus instance used in the control plane
+	PrometheusImage                 = "prom/prometheus:v2.7.1"
+	prometheusProxyOutboundCapacity = 10000
+	// DefaultDockerRegistry is the default registry to pull the proxy and init container images from
+	DefaultDockerRegistry = "gcr.io/linkerd-io"
+	// DefaultKeepaliveMs is used in the proxy configuration for remote connections
+	DefaultKeepaliveMs = 10000
+)
+
+var injectableKinds = []string{"Deployment", "ReplicationController",
+	"ReplicaSet", "Job", "DaemonSet", "StatefulSet", "Pod"}
+
+// objMeta provides a generic struct to parse the names of Kubernetes objects
+type objMeta struct {
+	*metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+}
+
+// ResourceConfig contains the parsed information for a given workload
+type ResourceConfig struct {
+	globalConfig    *config.Global
+	proxyConfig     *config.Proxy
+	nsAnnotations   map[string]string
+	meta            metav1.TypeMeta
+	obj             interface{}
```

Comment:
My intuition told me there had to be a higher abstraction for all the workload types :+1: 
Any particular reason though why not using the type switch? It alleviates us from doing that casting. In this case it would have to be pointers like `case *v1.Pod:`

Reply:
Ahh, yep, I couldn't get the type switch working last night, but now I realize it's because I was using object types instead of pointer types. Will push this change to that branch momentarily.